recursion/src/air/ext_alu_air.rs

```diff
@@ -0,0 +1,543 @@
+
+impl<T: Add<Output = T> + Clone> Add for BinomialExtension<T> {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(core::array::from_fn(|i| {
+            self.0[i].clone() + rhs.0[i].clone()
+        }))
+    }
+}
+
+impl<T: Sub<Output = T> + Clone> Sub for BinomialExtension<T> {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self(core::array::from_fn(|i| {
+            self.0[i].clone() - rhs.0[i].clone()
+        }))
+    }
+}
+
+impl<F: Add<Output = F> + Mul<Output = F> + Algebra<F>> Mul for BinomialExtension<F> {
+    type Output = Self;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        let mut result = [F::ZERO, F::ZERO, F::ZERO, F::ZERO];
+        // This value is specific for BabyBear prime's extension `F_p[x]/(x^4 - 11)`.
+        let w = F::from_u8(11); // TODO: Is this correct?
+
+        for i in 0..D {
+            for j in 0..D {
+                if i + j >= D {
+                    result[i + j - D] = result[i + j - D].clone()
+                        + w.clone() * self.0[i].clone() * rhs.0[j].clone();
+                } else {
+                    result[i + j] = result[i + j].clone() + self.0[i].clone() * rhs.0[j].clone();
+                }
+            }
+        }
+
+        Self(result)
+    }
+}
+
+impl<F> Distribution<BinomialExtension<F>> for StandardUniform
+where
+    StandardUniform: Distribution<F>,
+{
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> BinomialExtension<F> {
+        BinomialExtension(core::array::from_fn(|_| rng.random()))
+    }
+}
+
+#[repr(C)]
+/// Represents the columns in the ALU trace.
+/// R counts how many `a * b = c` operations to do per row in the AIR
+pub struct ExtAluCols<F, const R: usize = 1> {
+    pub left_addr: [F; R],
+    pub left_val: [BinomialExtension<F>; R],
+    pub right_addr: [F; R],
+    pub right_val: [BinomialExtension<F>; R],
+    pub res_addr: [F; R],
+    pub res_val: [BinomialExtension<F>; R],
+}
+
+impl<F, const R: usize> ExtAluCols<F, R> {
+    pub const TRACE_WIDTH: usize = 3 * R * (D + 1);
+}
+
+impl<F, const R: usize> Borrow<ExtAluCols<F, R>> for [F] {
+    fn borrow(&self) -> &ExtAluCols<F, R> {
+        debug_assert_eq!(self.len(), ExtAluCols::<F, R>::TRACE_WIDTH);
+        let (prefix, shorts, _suffix) = unsafe { self.align_to::<ExtAluCols<F, R>>() };
+        debug_assert!(prefix.is_empty(), "Alignment should match");
+        debug_assert_eq!(shorts.len(), 1);
+        &shorts[0]
+    }
+}
+
+impl<F, const R: usize> BorrowMut<ExtAluCols<F, R>> for [F] {
+    fn borrow_mut(&mut self) -> &mut ExtAluCols<F, R> {
+        debug_assert_eq!(self.len(), ExtAluCols::<F, R>::TRACE_WIDTH);
+        let (prefix, shorts, _suffix) = unsafe { self.align_to_mut::<ExtAluCols<F, R>>() };
+        debug_assert!(prefix.is_empty(), "Alignment should match");
+        debug_assert_eq!(shorts.len(), 1);
+        &mut shorts[0]
+    }
+}
+
+pub enum Operation {
+    Add,
+    Sub,
+    Mul,
+}
```

Comment:
There's already an `Operation` trait so it may be confusing to keep the same naming here. Maybe rename to `OpKind`?